server/src/analyser.ts

```diff
@@ -1,207 +1,214 @@
-
-      namedDeclarations.push(
-        SymbolInformation.create(name, kinds[n.type], range(named), uri),
-      )
-      declarations[uri][name] = namedDeclarations
+export class Analyzer {
+  /**
+   * Initialize the Analyzer based on a connection to the client and an optional
+   * root path.
+   *
+   * If the rootPath is provided it will initialize all *.sh files it can find
+   * anywhere on that path.
+   */
+  public static fromRoot(
+    connection: LSP.Connection,
+    rootPath: string | null,
+  ): Promise<Analyzer> {
+    // This happens if the users opens a single bash script without having the
+    // 'window' associated with a specific project.
+    if (!rootPath) {
+      return Promise.resolve(new Analyzer())
     }
-  })
 
-  return problems
-}
-
-/**
- * Find all the locations where something named name has been defined.
- */
-export function findDefinition(name: string): Location[] {
-  const symbols: SymbolInformation[] = []
-  Object.keys(declarations).forEach(uri => {
-    const declarationNames = declarations[uri][name] || []
-    declarationNames.forEach(d => symbols.push(d))
-  })
-  return symbols.map(s => s.location)
-}
+    return new Promise((resolve, reject) => {
+      glob('**/*.sh', { cwd: rootPath }, (err, paths) => {
+        if (err != null) {
+          reject(err)
+        } else {
+          const analyzer = new Analyzer()
+          paths.forEach(p => {
+            const absolute = Path.join(rootPath, p)
+            const uri = 'file://' + absolute
+            connection.console.log('Analyzing ' + uri)
+            analyzer.analyze(uri, fs.readFileSync(absolute, 'utf8'))
+          })
+          resolve(analyzer)
+        }
+      })
+    })
+  }
 
-/**
- * Find all the locations where something named name has been defined.
- */
-export function findReferences(name: string): Location[] {
-  const locations = []
-  Object.keys(documents).forEach(uri => {
-    findOccurrences(uri, name).forEach(l => locations.push(l))
-  })
-  return locations
-}
+  // Global map from uri to the tree-sitter document.
+  private documents: Documents = {}
 
-/**
- * Find all occurrences of name in the given file.
- * It's currently not scope-aware.
- */
-export function findOccurrences(uri: string, query: string): Location[] {
-  const doc = documents[uri]
-  const contents = texts[uri]
-
-  const locations = []
-
-  forEach(doc.rootNode, n => {
-    let name: string = null
-    let rng: Range = null
-
-    if (isReference(n)) {
-      const node = n.firstNamedChild || n
-      name = contents.slice(node.startIndex, node.endIndex)
-      rng = range(node)
-    } else if (isDefinition(n)) {
-      const namedNode = n.firstNamedChild
-      name = contents.slice(namedNode.startIndex, namedNode.endIndex)
-      rng = range(n.firstNamedChild)
-    }
+  // Global mapping from uri to the contents of the file at that uri.
+  // We need this to find the word at a given point etc.
+  private texts: Texts = {}
```

Comment:
We would rename all of these, so the comment would become superfluous.

Example: `uriToFileContent`, `uriToTreeSitterDocument`. I would prefer that. : )